pkg/operator/controller/controller.go

```diff
@@ -289,42 +195,68 @@ func (r *reconciler) ensureRouterForIngress(ci *ingressv1alpha1.ClusterIngress)
 		}
 	}
 
-	if err := r.ensureInternalRouterServiceForIngress(current, ci); err != nil {
+	if err := r.ensureRouterServiceInternal(deployment, ci); err != nil {
 		return fmt.Errorf("failed to create internal router service for clusteringress %s: %v", ci.Name, err)
 	}
 
 	return nil
 }
 
-// ensureInternalRouterServiceForIngress ensures that an internal service exists
-// for a given ClusterIngress.
-func (r *reconciler) ensureInternalRouterServiceForIngress(deployment *appsv1.Deployment, ci *ingressv1alpha1.ClusterIngress) error {
-	svc, err := r.ManifestFactory.RouterServiceInternal(ci)
-	if err != nil {
-		return fmt.Errorf("failed to build router service: %v", err)
-	}
-	err = r.Client.Get(context.TODO(), types.NamespacedName{Namespace: svc.Namespace, Name: svc.Name}, svc)
+// ensureRouterAsset ensures the expected router asset exists and returns the
+// current asset information.
+func (r *reconciler) ensureRouterAsset(key types.NamespacedName, obj runtime.Object) (runtime.Object, error) {
+	err := r.Client.Get(context.TODO(), key, obj)
```

Comment:
This is mutating the input which is surprising (no docs). Seems like a generic dynamic client concern, not something scoped to just router assets. Maybe extend the `Client` interface or something?

Also your deep copy additions reminded me that we're probably not being very safe generally in terms of mutating potentially cached items from our client rather than doing copies everywhere...